jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java

```diff
@@ -0,0 +1,283 @@
+
+    /**
+     * Construct the object for searching k shortest paths in provided graph using the provided
+     * shortest path algorithm.
+     * <p>
+     * It will immediately look for the overall shortest path using an instance of the algorithm
+     * provided by spFactory.
+     * 
+     * @param graph the underlying graph on which the paths are searched
+     * @param source the source/from/start vertex
+     * @param sink the sink/to/end vertex
+     * @param spFactory the factory function to create new instances of the shortest path algorithm.
+     */
+    public YenKShortestPathsIterator(
+        Graph<V, E> graph, V source, V sink,
+        Function<Graph<V, E>, ShortestPathAlgorithm<V, E>> spFactory)
+    {
+        this.spFactory = spFactory;
+        this.graph = graph;
+        this.sink = sink;
+        a = new ArrayList<GraphPath<V, E>>();
+        // Determine the shortest path from the source to the sink.
+        ShortestPathAlgorithm<V, E> shortestPathAlgorithm = spFactory.apply(graph);
+        GraphPath<V, E> p = shortestPathAlgorithm.getPath(source, sink);
+        if (null != p) {
+            a.add(p);
+        }
+        nextIndex = 0;
+        // // Initialize the heap to store the potential kth shortest paths.
+        b = new PriorityQueue<GraphPath<V, E>>(new Comparator<GraphPath<V, E>>()
+        {
+            @Override
+            public int compare(GraphPath<V, E> arg0, GraphPath<V, E> arg1)
+            {
+                return Double.compare(arg0.getWeight(), arg1.getWeight());
+            }
+        });
+    }
+
+    /**
+     * Construct the object for searching k shortest paths in provided graph using
+     * {@link DijkstraShortestPath} as the base algorithm.
+     * 
+     * @param graph the underlying graph on which the paths are searched
+     * @param source the source/from/start vertex
+     * @param sink the sink/to/end vertex
+     */
+    public YenKShortestPathsIterator(Graph<V, E> graph, V source, V sink)
+    {
+        this(graph, source, sink, DijkstraShortestPath<V, E>::new);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (nextIndex < a.size()) {
+            return true;
+        }
+        tryAddNext();
+        return nextIndex < a.size();
+    }
+
+    @Override
+    public GraphPath<V, E> next()
+    {
+        if (nextIndex < a.size()) {
+            return a.get(nextIndex++);
+        }
+        tryAddNext();
+        if (nextIndex >= a.size()) {
+            throw new NoSuchElementException();
+        }
+        return a.get(nextIndex++);
+    }
+
+    /**
+     * Determine and store in {@link #b} the next shortest paths and determine the actual shortest
+     * path and add it to {@link #a}. Instead of altering the underlying graph, this implementation
+     * masks edges and nodes that shall not be considered using {@link MaskSubgraph}.
+     */
+    private void tryAddNext()
+    {
+        GraphPath<V, E> lastShortestPath = a.get(a.size() - 1);

```

Comment:
This throws an exception if no paths are found.

Reply:
thx. indeed the most obvious case of no path was not handled properly. fixed.